examples/transformer/model.py

```diff
@@ -0,0 +1,195 @@
+from typing import Optional
+
+import torch
+from torch import nn
+
+import texar as tx
+
+
+class Transformer(nn.Module):
+    r"""A standalone sequence-to-sequence Transformer model.
+    TODO: Add detailed docstrings.
+    """
+
+    def __init__(self, model_config, data_config, vocab: tx.data.Vocab):
+        super().__init__()
+
+        self.config_model = model_config
+        self.config_data = data_config
+        self.vocab = vocab
+        self.vocab_size = vocab.size
+
+        self.word_embedder = tx.modules.WordEmbedder(
+            vocab_size=self.vocab_size,
+            hparams=self.config_model.emb,
+        )
+        self.pos_embedder = tx.modules.SinusoidsPositionEmbedder(
+            position_size=self.config_data.max_decoding_length,
+            hparams=self.config_model.position_embedder_hparams,
+        )
+
+        self.encoder = tx.modules.TransformerEncoder(
+            hparams=self.config_model.encoder
+        )
+        self.decoder = tx.modules.TransformerDecoder(
+            vocab_size=self.vocab_size,
+            output_layer=self.word_embedder.embedding,
+            hparams=self.config_model.decoder,
+        )
+
+        self.smoothed_loss_func = LabelSmoothingLoss(
+            label_confidence=self.config_model.loss_label_confidence,
+            tgt_vocab_size=self.vocab_size,
+            ignore_index=0,
+        )
+
+    def forward(  # type: ignore
+        self,
+        encoder_input: torch.Tensor,
+        is_train_mode: Optional[bool],
```

Comment:
`is_train_mode` differs from other examples where `mode` is used. Which one is preferred? Let's make them consistent

Reply:
and `mode` is usually the last argument

Reply:
Actually I think we don't need this variable. If the user provided `decoder_input`, then it's training mode; otherwise it's inference mode. It's also not a good practice to call `self.train()` and `self.eval()` inside `forward`.